src/analysis/processing/qgsalgorithmrandompointsonlines.cpp

```diff
@@ -0,0 +1,279 @@
+
+  //initialize random engine
+  srand( mRandSeed );
+
+  //index for finding close points (mMinDistance > 0)
+  QgsSpatialIndex index;
+
+  int totNPoints = 0;
+  int missedPoints = 0;
+  int missedLines = 0;
+  int emptyOrNullGeom = 0;
+  long tries = 0;
+  long saved = 0;
+
+  long featureCount = 0;
+  long numberOfFeatures = lineSource->featureCount();
+  QgsFeature lFeat;
+  QgsFeatureIterator fitL = mIncludeLineAttr ? lineSource->getFeatures()
+                            : lineSource->getFeatures( QgsFeatureRequest().setNoAttributes() );
+  while ( fitL.nextFeature( lFeat ) )
+  {
+    featureCount++;
+    if ( feedback->isCanceled() )
+    {
+      feedback->pushInfo( "canceled" );
+      break;
+    }
+    if ( !lFeat.hasGeometry() )
+    {
+      // Increment invalid features count
+      emptyOrNullGeom++;
+      continue;
+    }
+    QgsGeometry lGeom( lFeat.geometry() );
+    if ( lGeom.isEmpty() )
+    {
+      // Increment invalid features count
+      emptyOrNullGeom++;
+      continue;
+    }
+    double lineLength = lGeom.length();
+    int pointsAddedForThisFeature = 0;
+    for ( long i = 0; i < mNumPoints; i++ )
+    {
+      if ( feedback->isCanceled() )
+      {
+        break;
+      }
+      // Try to add a point (mMaxAttempts attempts)
+      int distCheckIterations = 0;
+      while ( distCheckIterations < mMaxAttempts )
+      {
+        if ( feedback->isCanceled() )
+        {
+          break;
+        }
+        distCheckIterations++;
+        tries++;
+        // Generate a random point
+        double randPos = lineLength * ( double ) rand() / RAND_MAX;
+        QgsGeometry rpGeom = QgsGeometry( lGeom.interpolate( randPos ) );
+
+        if ( !rpGeom.isNull() && !rpGeom.isEmpty() )
+        {
+          if ( mMinDistance != 0 && pointsAddedForThisFeature > 0 )
+          {
+            // Have to check minimum distance to existing points
+            QList<QgsFeatureId> neighbors = index.nearestNeighbor( rpGeom, 1, mMinDistance );
+            if ( !neighbors.empty() )
+            {
+              // Too close!
+              continue;
+            }
+          }
+          // point OK to add
+          QgsFeature f = QgsFeature( totNPoints );
+          QgsAttributes pAttrs = QgsAttributes();
+          pAttrs.append( totNPoints );
+          if ( mIncludeLineAttr )
+          {
+            pAttrs.append( lFeat.attributes() );
+          }
+          f.setAttributes( pAttrs );
+          f.setGeometry( rpGeom );
+
+          if ( mMinDistance != 0 )
+          {
+            index.addFeature( f );
+          }
+          sink->addFeature( f, QgsFeatureSink::FastInsert );
+          totNPoints++;
+          pointsAddedForThisFeature++;
+          saved += ( mMaxAttempts - distCheckIterations );
+
+          feedback->setProgress( static_cast<int>( 100 * static_cast<double>( tries + saved ) / static_cast<double>( mNumPoints * mMaxAttempts * ( numberOfFeatures - emptyOrNullGeom ) ) ) );
```

Comment:
Is this working correctly? I can't follow the logic here

Reply:
Sorry about that.

The reason why this one is so complicated is that I wanted to allow progress reporting also "within feature". So that it will be meaningful also for users that add an awful number of points on rather few line features.
The logic is:
- Denominator: The maximum number of tries that can be made:
  (*Number of features in the input line layer* - *Number of features without valid geometries*) * *Number of points per feature* * *Maximum number of attempts per feature*

- Nominator: The number of actual attempts + The number of "saved" attempts.
  When the minimum distance (`mMinDistance`) is 0, only one attempt will be made per point, and the rest of the attempts will be "saved". When the minimum distance is greater than 0, I make sure that the sum: *attempts* + *saved* is the maximum number of attempts per point (`mMaxAttempts`)
An attempt is saved when a point is added before the maximum number of attempts have been made

Thanks again for your patience!

Reply:
Ok - I get where you're going here! If it's ok with you'll I'll simplify in a follow up commit (without breaking your logic!) -- I'd like to merge this now :D